features/crc/soft/HalfByteCRC.h

```diff
@@ -0,0 +1,150 @@
+    HalfByteCRC(uint32_t polynomial = 0xEDB88320, uint32_t inital_xor = ~0x0, uint32_t final_xor = ~0x0)
+                : _polynomial(polynomial), _inital_value(inital_xor), _final_xor(final_xor)
+    {
+    }
+
+    virtual ~HalfByteCRC()
+    {
+    }
+
+    /** Initialize a CRC module, generate CRC static table
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    virtual int32_t init()
+    {
+        for (uint8_t i = 0; i < 16; i++) {
+            uint32_t crc = i * 16;
+            for (uint8_t j = 0; j < 8; j++) {
+                crc = (crc >> 1) ^ (-int(crc & 1) & _polynomial);
+            }
+            _crc_table[i] = crc;
+        }
+        return 0;
+    }
+
+    /** Deinitialize a CRC module
+     *
+     * @return  0 on success, negative error code on failure
+     */
+    virtual int32_t deinit()
+    {
+        return 0;
+    }
+
+    /** Get CRC polynomial
+     *
+     * Get the current CRC polynomial
+     *
+     * @return  Polynomial value
+     */
+    virtual uint32_t get_polynomial(void) const
+    {
+        return _polynomial;
+    }
+
+    /** Compute CRC for the data input as per the polynomial set
+     *
+     *  @param  buffer  Data bytes
+     *  @param  size  Size of data
+     *  @param  crc  CRC result
+     *  @return  0  on success or a negative error code on failure
+     *  @note CRC calculations are done byte wise.
+     */
+    virtual int32_t compute(void *buffer, crc_data_size_t size, uint32_t *crc)
+    {
+        MBED_ASSERT(crc != NULL);
+        MBED_ASSERT(buffer != NULL);
+
+        uint8_t *data = static_cast<uint8_t *>(buffer);
+        uint32_t remainder = _inital_value;
+
+        for (crc_data_size_t byte = 0; byte < size; byte++) {
+            remainder = _crc_table[((remainder ^ (data[byte] >> 0)) & 0x0F)] ^ (remainder >> 4);
+            remainder = _crc_table[((remainder ^ (data[byte] >> 4)) & 0x0F)] ^ (remainder >> 4);
+        }
+
+        *crc = (remainder ^ _final_xor);
+        return 0;
+    }
+
+    /** Compute partial CRC for the data input.
+     *
+     *  CRC data if not available fully, CRC can be computed in parts with available data.
+     *  Previous CRC as an argument to function should be the CRC output of previous
+     *  computation, in case of first partial computation it should be set as 0x0.
+     *
+     *  @param  buffer  Data bytes
+     *  @param  size  Size of data
+     *  @param  p_crc  Previous CRC to continue CRC computation, should be 0 for first part
+     *  @param  crc  CRC result
+     *  @return  0  on success or a negative error code on failure
+     *  @note CRC calculations are done byte wise.
+     */
+    virtual int32_t compute_partial(void *buffer, crc_data_size_t size, uint32_t p_crc, uint32_t *crc)
+    {
+        MBED_ASSERT(crc != NULL);
+        MBED_ASSERT(buffer != NULL);
+
+        uint8_t *data = static_cast<uint8_t *>(buffer);
+        uint32_t remainder = p_crc == 0x0 ? _inital_value : ~p_crc;
```

Comment:
p_crc == 0x0 might be a valid intermediary value for the crc, this would break if that lined up with a `compute_partial`  block. You can expect the user to either initialize it themselves, or call `compute` first

Reply:
Asking user to set initial value will be better then calling compute first. Will update this.